OpenRA.Mods.Common/Traits/Buildings/Gate.cs

```diff
@@ -0,0 +1,144 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public interface ITemporaryBlocker
+	{
+		bool CanLiftBlockade(Actor self, Actor blocking);
+		bool Blocks();
+	}
+
+	[Desc("Will open and be passable for actors that appear friendly when there are no enemies in range.")]
+	public class GateInfo : BuildingInfo, ITraitInfo, Requires<WithSpriteBodyInfo>
+	{
+		public readonly string OpenSequence = "open";
+		public readonly string ClosingSequence = "closing";
+		public readonly string ClosedSequence = "closed";
+
+		public readonly string OpeningSound = "gateup1.aud";
+		public readonly string ClosingSound = "gatedwn1.aud";
+
+		[Desc("How many ticks until the gate will close after being opened.")]
+		public readonly int TicksToStayOpen = 50;
+
+		[Desc("How many ticks until the gate is open.")]
+		public readonly int TicksToOpen = 10;
+
+		public override object Create(ActorInitializer init) { return new Gate(init, this); }
+	}
+
+	public class Gate : Building, ITick, ITemporaryBlocker, INotifyBlockingMove
+	{
+		readonly GateInfo info;
+		readonly Actor self;
+		readonly WithSpriteBody wsb;
+
+		bool isOpen;
+		bool opening;
+
+		int remainingTime;
+
+		public Gate(ActorInitializer init, GateInfo info) : base(init, info)
+		{
+			this.info = info;
+			this.self = init.Self;
+
+			wsb = self.Trait<WithSpriteBody>();
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (self.IsDisabled() || !BuildComplete)
+				return;
+
+			if (opening && remainingTime-- <= 0)
+				Opened();
+			else if (isOpen)
+			{
+				if (remainingTime-- <= 0)
+				{
+					if (!IsBlocked())
+						Close();
+					else
+						remainingTime = info.TicksToStayOpen;
+				}
+				else if (remainingTime % 10 == 0 && IsBlocked())
+					remainingTime = info.TicksToStayOpen;
+			}
+		}
+
+		bool ITemporaryBlocker.Blocks()
+		{
+			return !isOpen;
+		}
+
+		bool ITemporaryBlocker.CanLiftBlockade(Actor self, Actor blocking)
+		{
+			return !self.IsDisabled() && BuildComplete && blocking.AppearsFriendlyTo(self);
+		}
+
+		void INotifyBlockingMove.OnNotifyBlockingMove(Actor self, Actor blocking)
+		{
+			if (!self.IsDisabled() && BuildComplete && !isOpen && (this as ITemporaryBlocker).CanLiftBlockade(self, blocking))
+				Open();
+		}
+
+		bool IsBlocked()
+		{
+			var eligibleLocations = FootprintUtils.Tiles(self).ToList();
```

Comment:
The `.ToList()` here should be unnecessary. You only enumerate it once in the `foreach`.